servicetalk-tcp-netty-internal/src/main/java/io/servicetalk/tcp/netty/internal/TcpConnector.java

```diff
@@ -209,4 +206,87 @@ private NoopNettyAddressResolverGroup() {
             return NOOP_ADDRESS_RESOLVER;
         }
     }
+
+    private static final class ConnectHandler<C extends ListenableAsyncCloseable> implements Consumer<Channel> {
+        private static final Logger LOGGER = LoggerFactory.getLogger(ConnectHandler.class);
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<ConnectHandler> terminatedUpdater =
+                newUpdater(ConnectHandler.class, "terminated");
+
+        private final DelayedCancellable futureCancellable = new DelayedCancellable();
+        private final DelayedCancellable flatMapCancellable = new DelayedCancellable();
+        private final Subscriber<? super C> target;
+        private final Function<Channel, Single<? extends C>> connectionFactory;
+
+        private volatile int terminated;
+
+        ConnectHandler(final Subscriber<? super C> target,
+                       final Function<Channel, Single<? extends C>> connectionFactory) {
+            this.target = target;
+            this.connectionFactory = connectionFactory;
+            target.onSubscribe(() -> {
+                try {
+                    futureCancellable.cancel();
+                } finally {
+                    flatMapCancellable.cancel();
+                }
+            });
+        }
+
+        @Override
+        public void accept(final Channel channel) {
+            toSource(connectionFactory.apply(channel)
+                    .subscribeShareContext())
+                    .subscribe(new Subscriber<C>() {
+                        @Override
+                        public void onSubscribe(final Cancellable cancellable) {
+                            flatMapCancellable.delayedCancellable(cancellable);
+                        }
+
+                        @Override
+                        public void onSuccess(final C connection) {
+                            requireNonNull(connection);
```

Comment:
`SingleSource.Subscriber` defines a parameter for `onSuccess` as `@Nullable`. Instead of throwing from here, should we propagate `onError` instead?

Reply:
Good point, the `null-check` is only required for when we want to call `close()` so instead of disallowing `null`, lemme guard the `close`.